lv_mem: Add interface for safe reallocation

C47D opened the pull request:
### Description of the feature or fix

Safe memory reallocation, fixes #2194 

### Checkpoints
- [x] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [x] Run `code-format.py` from the `scripts` folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the documentation

MaJerle reviewed (commented):
`bool lv_mem_realloc_safe(void * data, size_t new_size);` should accept rather `void ** data`

MaJerle reviewed (commented):
You need *data = new_ptr.

Has this been tested in reality?

C47D commented:
@MaJerle tested only compiling the source code. I guess I can write an unit test that reallocates a bunch of memory until it fails.

MaJerle commented:
> @MaJerle tested only by compiling the source code. I guess I can write an unit test that reallocates a bunch of memory until it fails.

A test with success and failure would be nice for this function. I see some potential wrong implementation.

MaJerle commented on tests/src/test_cases/test_mem.c:
```diff
@@ -0,0 +1,44 @@
+
+    void * ptr = NULL;
+    void * data = &ptr;
+    retval = lv_mem_realloc_safe(data, allocation_size);
```
Are you sure this works well? Don't you need to pass &data to realloc?

MaJerle commented on src/misc/lv_mem.c:
```diff
@@ -210,22 +210,31 @@ bool lv_mem_realloc_safe(void ** data, size_t new_size)
 
-    return new_p != NULL;
+    MEM_TRACE("Memory allocated at %p", new_p);
+    data = new_p;
```
Should be rather `*data = new_p`.

kisvegabor commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +204,39 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+
+    MEM_TRACE("reallocating %p with %lu size", *data, (unsigned long)new_size);
+
+    if((*data == NULL) && (0 < new_size)) {
```
`size_t` is always unsigned so it can't be negative.
https://stackoverflow.com/questions/1089176/is-size-t-always-unsigned

C47D replied:
Will replace with a `0 != new_size` check

MaJerle replied:
It cant be negative but can be zero. Check seems legit unless there are other code style rules by lib.

kisvegabor commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +204,39 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+
+    if((*data == NULL) && (0 < new_size)) {
+        *data = lv_mem_alloc(new_size);
+        return false;
```
What about returning `LV_RES_OK/INV`? It's cleaner then `true/false`.

C47D replied:
It's ok, I was thinking on returning the status of the allocation. `LV_RES_OK` if the allocation was done, `LV_RES_INV` otherwise. Are those enums already available?

C47D replied:
Resolved with 48c3333

kisvegabor commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +204,39 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+        *data = lv_mem_alloc(new_size);
+        return false;
+    }
+    else if((*data != NULL) && (0 == new_size)) {
+        MEM_TRACE("using zero_mem");
+        lv_mem_free(*data);
+        *data = &zero_mem;
+        return false;
+    }
+
+#if LV_MEM_CUSTOM == 0
+    new_p = lv_tlsf_realloc(tlsf, data, new_size);
+#else
+    new_p = LV_MEM_CUSTOM_REALLOC(data, new_size);
+#endif
+    if(new_p == NULL) {
+        LV_LOG_ERROR("couldn't allocate memory");
+        return false;
+    }
```
All these should be handled in `lv_mem_realloc()`

MaJerle replied:
Maybe have a look here and please read func desc. Implementation of safe version shall be super super non-complex. 

https://github.com/MaJerle/lwmem/blob/de2917481ddcb02d4789b188f7c8209af6cdcc17/lwmem/src/lwmem/lwmem.c#L894

C47D commented:
It seems to me that I don't fully understand double pointers, I can't get the test to pass, I think I leaved the prototype documented similar to the reference implementation. I will leave this in your hands @kisvegabor @MaJerle

MaJerle commented on tests/src/test_cases/test_mem.c:
```diff
@@ -0,0 +1,40 @@
+/* Non-zero size with null pointer should behave like malloc */
+void test_realloc_safe_behave_malloc_like(void)
+{
+    void * ptr = NULL;
```
You need only one variable, either ptr or data, but not both.

C47D replied:
Thanks, fixed in 986a62d

MaJerle commented on tests/src/test_cases/test_mem.c:
```diff
@@ -0,0 +1,40 @@
+    void * data = &ptr;
+    size_t allocation_size = sizeof(lv_font_t);
+
+    lv_res_t retval = lv_mem_realloc_safe(&data, allocation_size);
```
This looks OK

MaJerle commented on tests/src/test_cases/test_mem.c:
```diff
@@ -0,0 +1,40 @@
+    TEST_ASSERT_EQUAL_UINT16(retval, LV_RES_OK);
+
+    /* Don't leak memory */
+    lv_mem_free(ptr);
```
You free either data or ptr, the variable you have, but not both.

C47D replied:
So it is ok to only free `ptr`?

MaJerle commented on tests/src/test_cases/test_mem.c:
```diff
@@ -0,0 +1,40 @@
+void test_realloc_safe_behave_free_like(void)
+{
+    void * font = lv_mem_alloc(1 * sizeof(lv_font_t));
+    void * data = &font;
```
Same here. proper test code should be

```c
void* font = lv_mem_alloc(1 * sizeof(lv_font_t));

//And then to realloc (or free in this case since len is 0)
lv_res_t retval = lv_mem_realloc_safe(&font, 0);

// font should now point to NULL as safe function should set it to NULL as it was free operation
if (font != NULL) {
    //Error
}
```

C47D replied:
I've refactored the test a bit, can you check 503e757?

C47D replied:
It currently sets the pointer to `zero_mem`, should it set it to NULL instead?

C47D replied:
How does 60b0f318 look? I think we could also move `ZERO_MEM_SENTINEL` to `lv_mem.h` to be able to use it on the tests, any thoughts @kisvegabor ?

MaJerle commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +205,36 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+    MEM_TRACE("Safely reallocating %p with %lu size", *data, (unsigned long) new_size);
+
+#if LV_MEM_CUSTOM == 0
+    new_p = lv_tlsf_realloc(tlsf, data, new_size);
```
You need existing address, so it must be `*data` in parameter

C47D replied:
Done in 503e757

MaJerle commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +205,36 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+#if LV_MEM_CUSTOM == 0
+    new_p = lv_tlsf_realloc(tlsf, data, new_size);
+#else
+    new_p = LV_MEM_CUSTOM_REALLOC(data, new_size);
```
You need existing address, so it must be `*data` in parameter

C47D replied:
Done in 503e757

MaJerle commented on src/misc/lv_mem.c:
```diff
@@ -204,6 +205,36 @@ void * lv_mem_realloc(void * data_p, size_t new_size)
+        return LV_RES_INV;
+    }
+
+    *data = new_p;
```
This looks good

C47D commented:
Thanks for the review @MaJerle , it seems like all I needed was a bit of guidance and go away from the PC for some hours. I've addressed most of your comments, only 2 are left opened for your feedback. What timezone are you in? I'm in Mexico.

MaJerle commented on src/misc/lv_mem.c:
```diff
@@ -182,13 +182,14 @@ void lv_mem_free(void * data)
     }
-
-    if(data_p == &zero_mem) return lv_mem_alloc(new_size);
+    if(data_p == &zero_mem && new_size != 0) {
```
We need to check if it is NULL (first try) or if it is zero mem.

`if((data_p == &zero_mem || data_p == NULL) && new_size != 0) {`

C47D replied:
Resolved in 0e0bebd

MaJerle commented:
I'm in EMEA, CET timezone, like Gabor.

C47D commented:
There's only one open question (I think). What is the plan, merging this and then work on using it through the code base in another PR?

kisvegabor commented:
It looked great in general. So far I haven't tried the "test-driven-development" and to be honest it feels great. Thank you, Carlos for  introducing this pattern in LVGL. :slightly_smiling_face:   

I've added a few minor changes to the code.  What do you think?

> What is the plan, merging this and then work on using it through the code base in another PR?

Adding the changes should be straightforward so I think we can add them in this PR.

X-Ryl669 commented on demos/README.md:
````diff
@@ -53,11 +53,20 @@
+For running this demo properly, please make sure **LV_MEM_SIZE** is at least **38KB** (and **48KB** is recommended):
+
+```c
+#define LV_MEME_SIZE    (38ul * 1024ul)
````
Seems like a typo.

C47D replied:
Good catch. It was introduced in 7f4cde9388a649fa31fb286dd1052d1108eb7029, will send the fix.

C47D replied:
See f93ce64

C47D commented:
@kisvegabor was the free-like test removed by accident?

C47D commented:
@kisvegabor @MaJerle Some time ago I used [fff](https://github.com/meekrosoft/fff) to fake functions in this project [WFlash](https://github.com/C47D/WFlash/blob/master/test/test_wflash.c), you can see the project also uses Unity, so they play along nicely. I think we could fake the behavior of `lv_mem_realloc` inside `lv_mem_realloc_safe` and test that the input pointer doesn't get updated, etc.

Also we can define what should happen when reallocation fails and also test it, e.g. what should happen when reallocation fails when calling `lv_checkbox_set_text`?

What do you think?

kisvegabor commented:
> was the free-like test removed by accident?

It is handled in `lv_mem_realloc` (i.e. freeing and returning `&zero_mem`)

> Can't define assert handlers in the tests CMakeList.txt file because CMake thinks we end the command with the assert parenthesis. Adding this to the option lists breaks it.

it'd be really great but from [here](https://cmake.org/cmake/help/latest/prop_dir/COMPILE_DEFINITIONS.html):
> Function-style definitions are not supported

But we can add something like this in `lv_test_conf.h`
```c
#if LV_TEST_CUSTOM_ASSERT
# define LV_ASSERT_HANDLER {extern int lv_assert_counter; lv_assert_counter++;}
#endif
```

C47D commented:
We can leave testing assertions for later (I think), is the change in checkbox the expected from introducing the new API?

C47D commented:
@kisvegabor should we continue with this PR?

kisvegabor commented:
Based on @X-Ryl669's comments in #2194 I'm not sure really this is the best direction. 

Implementing @xiaoxiang781216's "emergency free hook" should solve more problems with less effort.

So I suggest closing this PR for now. We can pick up from here if we really want to add a safe realloc later.

What do you  think?

C47D commented:
It's ok, I think we are solving two different issues. I will check the information on the comment.
